### PR TITLE
Raises exception when using coverage with clang

### PR DESCRIPTION
**Pull request by mayl**
Per #6160, it seems like meson does not actually support coverage builds with clang.  Currently, it fails in an unexpected way.  This causes it to raise an exception at configuration time, indicating to the user that clang is not supported with coverage builds.

Opening this as draft because the only way I found to check if meson is using clang is really janky.  What would be the "correct" way to do this?  Am I on the right path raising an exception here?  

I also understand I'll need to provide a test for this.  Currently, I'm using `meson init` to make a simple one test C project and the building with `CC=clang` and `-Db_coverage=true`.  Should I be adding that to the tests, or would we prefer just a unit test that proves the exception is raised?

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -320,6 +320,9 @@ def generate(self, interp):
             self.generate_dist()
             if 'b_coverage' in self.environment.coredata.base_options and \
                     self.environment.coredata.base_options['b_coverage'].value:
+                if 'clang' in repr(self.environment.coredata.compilers[MachineChoice['BUILD']]['c']):
+                    raise mesonlib.EnvironmentException(
+                            'Meson does not support coverage builds with Clang')
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[Flake8]__ continuation line over-indented for hanging indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6192#issue-4846559))

<details>

| Rule |
| - |
| E126 |

</details>
<sub>posted by Sider</sub>

**Comment by scivision** [comment]
I didn't have a chance to try it yet, but did the proposed fix in #6188 make coverage work for you?

**Comment by mayl** [comment]
Hey, thanks for the comment. I haven't tried that fix, but it doesn't seem to apply.  I am not on macOS and this is not a new problem for Clang on Linux for me.  It's just coincidental that I'm getting around to submitting these issues at the same time as the coverage regression on macOS. I'm actually surprised it actually works on macOS.

**Comment by scivision** [comment]
OK thanks

**Comment by krader1961** [comment]
> Hey, thanks for the comment. I haven't tried that fix, but it doesn't seem to apply. I am not on macOS and this is not a new problem for Clang on Linux for me. It's just coincidental that I'm getting around to submitting these issues at the same time as the coverage regression on macOS. I'm actually surprised it historically works on macOS at all.

That fix works for me on macOS with clang. You should be able to apply the fix even if you are not on macOS. It historically worked on macOS because the Xcode compiler, clang, was deemed "C like" and thus included the `--coverage` flag. If that change/fix doesn't apply cleanly on your system that implies you are not using the current master branch head.

**Comment by jpakkane** [comment]
Closing as per the discussion in #6197.

**Pull request closed without merging**